### PR TITLE
New hood rollers pid subsystem

### LINKED ISSUE #123: Program NEO for hood

### PR DESCRIPTION
**Pull request by agrinmanriv0537**
Closes #123

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/ShooterRoller.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    @Override
+    public void periodic() {
+        if (m_enabled) {
+            useOutput(m_controller.calculate(getMeasurement(), getSetpoint()), getSetpoint());
+
+            // double selSenVel = shooter.getSelectedSensorVelocity(0);
+            // double rotPerSec = (double) selSenVel / Constants.ShooterPID.kUnitsPerRevolution
+            // * 10; /* scale per100ms to perSecond */
+
+            // System.out.println("SHOOTER RPM (Speed): " + rotPerSec * 60);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/ShooterRoller.java]: remove. Its not correct anyways

File: src/main/java/frc/robot/subsystems/ShooterRoller.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public double getMeasurement() {
+        RelativeEncoder encoder =
+            shooterRoller.getEncoder(SparkMaxRelativeEncoder.Type.kHallSensor, 42);
+        encoder.setVelocityConversionFactor(0.5);
+        double selSenVel = encoder.getVelocity();
+        // double selSenVel = shooterRoller.getEncoder(SparkMaxRelativeEncoder.Type.kHallSensor,
+        // 42);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/ShooterRoller.java]: Whats this for

File: src/main/java/frc/robot/subsystems/ShooterRoller.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        shooterRoller.set(.3);
+    }
+
+    /**
+     * Stops the shooter roller.
+     */
+    public void stopShooterRoller() {
+        shooterRoller.set(0);
+    }
+
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/ShooterRoller.java]: What is this for??

File: src/main/java/frc/robot/subsystems/ShooterRoller.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    public ShooterRoller() {
+        super(new PIDController(Constants.ShooterRollerPID.kP, Constants.ShooterRollerPID.kI,
+            Constants.ShooterRollerPID.kD));
+        shooterRoller.setInverted(true); // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+        // shooterRoller.configSelectedFeedbackSensor(FeedbackDevice.IntegratedSensor, 1, 1);
+        getController().setTolerance(Constants.ShooterRollerPID.kShooterRollerToleranceRPS);
+        // IN
+        // RPS
+        // NOT
+        // RPM
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/ShooterRoller.java]: WHy

File: src/main/java/frc/robot/subsystems/ShooterRoller.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    private final SimpleMotorFeedforward shooterRollerFeed = new SimpleMotorFeedforward(
+        Constants.ShooterRollerPID.kSVolts, Constants.ShooterRollerPID.kVVoltSecondsPerRotation);
+
+    /**
+     * Create shooter roller class for PID
+     */
+    public ShooterRoller() {
+        super(new PIDController(Constants.ShooterRollerPID.kP, Constants.ShooterRollerPID.kI,
+            Constants.ShooterRollerPID.kD));
+        shooterRoller.setInverted(true); // !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (legoguy1000) [inline src/main/java/frc/robot/subsystems/ShooterRoller.java]: WTF is this `// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!`

File: src/main/java/frc/robot/autos/P_2B.java [hunk 6]
```diff
@@ -48,16 +51,19 @@ public P_2B(Swerve swerve, Shooter shooter, Magazine magazine, Intake intake, Tu
         addCommands(new InstantCommand(() -> swerve.resetOdometry(trajectory.getInitialPose())),
             new InstantCommand(() -> turret.alignEnabled = true),
             new InstantCommand(() -> this.shooter.setSetpoint(4500 / 60)),
+            new InstantCommand(() -> this.shooterRoller.setSetpoint((4500 / 60))),
```
Comment (legoguy1000) [inline src/main/java/frc/robot/autos/P_2B.java]: Should be double

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -155,25 +157,31 @@ private void configureButtonBindings() {
...
                     new MagazineRPM(shooter, magazine))))
             .whenReleased(new InstantCommand(shooter::disable, shooter))
+            .whenReleased(new InstantCommand(shooterRoller::disable, shooterRoller))
             .whenReleased(new InstantCommand(magazine::disable, magazine));
 
         new AxisButton(operator, XboxController.Axis.kRightTrigger.value)
             .whileHeld(new SequentialCommandGroup(
                 new InstantCommand(() -> shooter.setSetpoint(4300 / 60), shooter),
                 new InstantCommand(() -> shooter.enable(), shooter),
+                new InstantCommand(() -> shooterRoller.setSetpoint(4300 / 60), shooterRoller),
```
Comment (legoguy1000) [inline src/main/java/frc/robot/RobotContainer.java]: Should be double.  Use a Variable and *2

File: src/main/java/frc/robot/autos/P_2B.java (hunk 6 above)
Comment (legoguy1000) [reply src/main/java/frc/robot/autos/P_2B.java]: Use a variable and *2

File: src/main/java/frc/robot/commands/ShooterRPM.java [hunk 8]
```diff
@@ -3,47 +3,75 @@
...
+     *
+     * @param shooter shooter subsystem
+     * @param shooterRoller shooter roller subsystem
+     * @param rps Hardcoded setpoint for Shooter
+     */
+    public ShooterRPM(Shooter shooter, ShooterRoller shooterRoller, double rps) {
+        this.shooter = shooter;
+        this.shooterRoller = shooterRoller;
+        this.setRPS = rps;
+        addRequirements(shooter, shooterRoller);
```
Comment (legoguy1000) [inline src/main/java/frc/robot/commands/ShooterRPM.java:43]: @agrinmanriv0537  This should work right???

**Review by aidund — APPROVED** [review]

**Pull request merged**